platform/macos/godot_main_macos.mm

```diff
@@ -74,6 +74,7 @@ int main(int argc, char **argv) {
 	}
 
 	if (Main::start()) {
+		os.set_exit_code(EXIT_SUCCESS);
```

Comment:
I'm not sure this is the right solution. See:

https://github.com/godotengine/godot/blob/fc99492d3066098e938449b10e02f8e01d07e2d1/core/os/os.h#L59

* godotengine/godot-proposals#7557
* #30207
* #38929

Reply:
I guess then #67421 is wrong as well?

Though looking at #38929 I do not quite follow what effect the default error value would have there, the comment mentions "unexpected exit" which I do not fully understand what is meant by that. If the application crashes, it would never reach the normal process return case anyway?

Maybe @touilleMan can clarify the reason of the error assigned on init and how this should be fixed differently?

Reply:
>  If the application crashes, it would never reach the normal process return case anyway?

The issue comes if Godot doesn't crash, but the game is doesn't load either (typically the main scene fails to load because a resource is missing).
Godot is designed around the idea of "keep going when something goes wrong" (from what I understand the idea is that it's better to have a game run with quirks than to have straight crash). But it our case this strategy means the failure to load the main scene end up with an error log and the game engine carries on, discover it has nothing more to do, and exit. So if the exit code is not initialized at an error value, we will end up with a success value returned here.

So I wouldn't be surprised #67421 reintroduced the issue solved by #38929

Also see my comment https://github.com/godotengine/godot/pull/38929#discussion_r435158647, so as long as `SceneTree.quit()` is called the status code must be correct (given it is provided in as parameter).
So my guess here is closing the game is done is some exotic way instead of relying on `SceneTree.quit()`, the "correct" to fix this issue would be to investigate this and correct this exotic way of close to rely on `SceneTree.quit()` (or to set the status code before exiting)

(It's been a long time I haven't work on this, so I might be totally wrong here :smile: )

Reply:
Just an early draft, but after discussing with @mihe who was looking into this, I decided to have a go at changing the default exit code to `EXIT_SUCCESS`: https://github.com/akien-mga/godot/commit/1ee4a050434bf5d6e7e07d139b98baec7a315d86

I notably changed `Main::start()` to return the exit code instead of a bool, which makes it much easier to keep track of whether we're returning early with a failure state (set exit code) or not (proceed further, exit code is success by default, or may be overridden by iteration failures).

Haven't really tested it yet, there might well be issues when the application crashes. @mihe and I will try to get to the bottom of this and finally aim to solve all these pesky issues.

Reply:
> So I wouldn't be surprised https://github.com/godotengine/godot/pull/67421 reintroduced the issue solved by https://github.com/godotengine/godot/pull/38929

Did a quick test with my approach, failing to loading a main scene seems to properly report 1 as exit code:
```
$ godot-git
Godot Engine v4.3.dev.custom_build.1ee4a0504 (2024-03-06 17:53:16 UTC) - https://godotengine.org
Vulkan 1.3.267 - Forward+ - Using Device #1: AMD - AMD Radeon RX 7600M XT (RADV NAVI33)

ERROR: Cannot open file 'res://Node2.tscn'.
   at: load (./scene/resources/resource_format_text.cpp:1656)
ERROR: Failed loading resource: res://Node2.tscn. Make sure resources have been imported by opening the project in the editor at least once.
   at: _load (./core/io/resource_loader.cpp:277)
ERROR: Failed loading scene: res://Node2.tscn.
   at: start (main/main.cpp:3790)
~/.../bug/crashing$ echo $?
1
```

Needs a lot more testing on all kinds of scenarios where things may fail. I guess I'll open a draft PR and move the discussion there.